File: androidApp/src/main/kotlin/com/mohamedfaridelsherbini/moventiq/ui/permissions/PermissionFlowStepResolver.kt
package com.mohamedfaridelsherbini.moventiq.ui.permissions

/**
 * Resolves which permission screen to show. Each step has its own skip rule —
 * location defer does not block notification re-prompt after location is granted.
 */
internal object PermissionFlowStepResolver {

    fun resolve(input: PermissionFlowInput): PermissionFlowStep =
        when {
            input.limitedFeaturesAcknowledged -> PermissionFlowStep.None
            input.hasAdequateLocationAccess -> resolveNotificationStep(input)
            input.showLocationDeniedRecovery -> PermissionFlowStep.Denied
            !input.locationSkippedThisSession -> PermissionFlowStep.Location
            else -> resolveNotificationStep(input)
        }

    private fun resolveNotificationStep(input: PermissionFlowInput): PermissionFlowStep =
        when {
            !input.notificationPromptRequired || input.notificationGranted -> PermissionFlowStep.None
            input.notificationSkippedThisSession -> PermissionFlowStep.None
            else -> PermissionFlowStep.Notification
        }
}

internal data class PermissionFlowInput(
    val limitedFeaturesAcknowledged: Boolean,
    val hasAdequateLocationAccess: Boolean,
    val showLocationDeniedRecovery: Boolean,
    val locationSkippedThisSession: Boolean,
    val notificationSkippedThisSession: Boolean,
    val notificationPromptRequired: Boolean,
    val notificationGranted: Boolean,
)

File: iosApp/iosApp/Features/Permissions/PermissionFlowStepResolver.swift
import Foundation

enum PermissionFlowStepResolver {
    static func resolve(_ input: PermissionFlowInput) -> PermissionFlowStep {
        if input.limitedFeaturesAcknowledged {
            return .none
        }

        if input.hasAdequateLocationAccess {
            return resolveNotificationStep(input)
        }

        if input.showLocationDeniedRecovery {
            return .denied
        }

        if !input.locationSkippedThisSession {
            return .location
        }

        return resolveNotificationStep(input)
    }

    private static func resolveNotificationStep(_ input: PermissionFlowInput) -> PermissionFlowStep {
        if !input.notificationPromptRequired || input.notificationGranted {
            return .none
        }
        if input.notificationSkippedThisSession {
            return .none
        }
        return .notification
    }
}

struct PermissionFlowInput: Equatable {
    let limitedFeaturesAcknowledged: Bool
    let hasAdequateLocationAccess: Bool
    let showLocationDeniedRecovery: Bool
    let locationSkippedThisSession: Bool
    let notificationSkippedThisSession: Bool
    let notificationPromptRequired: Bool
    let notificationGranted: Bool
}
